### PR TITLE
bpo-34622: Extract asyncio exceptions into a separate module

### PR DESCRIPTION
**Pull request by asvetlov**
https://bugs.python.org/issue34622

### CODE REVIEW

File: Lib/asyncio/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+__all__ = ('CancelledError', 'InvalidStateError', 'TimeoutError',
+           'IncompleteReadError', 'LimitOverrunError',
+           'SendfileNotAvailableError')
+
+import concurrent.futures
+from . import base_futures
+
+
+class CancelledError(concurrent.futures.CancelledError):
```
Comment (serhiy-storchaka) [inline Lib/asyncio/exceptions.py:12]: Why define subclasses instead of just making aliases?

**Review by serhiy-storchaka — COMMENTED** [review]
You need to set `__module__ = 'asyncio'` in all classes for pickle compatibility. This is needed if you want to unpickle pickles created in 3.8 in older Python versions.

Or set fake `__name__ = 'asyncio'` in the asyncio.exceptions module.

I don't know whether it is a good idea to move all exceptions.

File: Lib/asyncio/exceptions.py (hunk 1 above)
Comment (1st1) [reply Lib/asyncio/exceptions.py:12]: The main motivation is to improve usability.  There's no fundamental reason why these exceptions are derived from `concurrent.future`, it happened pretty much by accident.  But the problem is that asyncio users see `concurrent.future.TimeoutError` exceptions occurring in their programs and have no idea why it's not `asyncio.TimeoutError`.

Moreover, we have a plan to make `asyncio.CancelledError` a `BaseException` in a follow up PR.

File: Lib/asyncio/futures.py [hunk 2]
```diff
@@ -288,6 +285,17 @@ def _set_result_unless_cancelled(fut, result):
     fut.set_result(result)
 
 
+def _convert_future_exc(exc):
+    if type(exc) is concurrent.futures.CancelledError:
```
Comment (1st1) [inline Lib/asyncio/futures.py]: Can you cache the `type(exc)` call?

**Review by 1st1 — APPROVED** [review]

File: Lib/asyncio/futures.py (hunk 2 above)
Comment (asvetlov) [reply Lib/asyncio/futures.py]: Done

**Comment by asvetlov** [comment]
Exception classes are not stored persistently.
Pickling an exception, e.g. for passing into process worker by `multiprocesses` call should work pretty fine.

We are creating new exception classes anyway, no need to use tricks for very rare and weird backward compatibility use cases.

A code from python < 3.8 will be unpickled into `concurrent.futures` exceptions. Not ideal but still not the world breakage.

**Comment by 1st1** [comment]
@asvetlov I agree, feel free to merge as soon as CI is green.

**Pull request merged**